File: tools/compute-new-world-embeddings.mjs
// ============================================================
// tools/compute-new-world-embeddings.mjs
// 为两个新世界（克苏鲁 + 都市怪谈）的 lore 条目计算 512 维向量
// 用法：
//   node tools/compute-new-world-embeddings.mjs
// ============================================================
import { pipeline, env } from "@xenova/transformers";
import { writeFile } from "node:fs/promises";
import { fileURLToPath, pathToFileURL } from "node:url";
import path from "node:path";

const _dirname = path.dirname(fileURLToPath(import.meta.url));
const _root = path.resolve(_dirname, "..");

// 使用本地模型（已量化，~24MB），不走远程下载
env.localModelPath = pathToFileURL(path.join(_root, "models")).href;
// 镜像
env.remoteHost = "https://hf-mirror.com/";

const EMBED_MODEL = "Xenova/bge-small-zh-v1.5";
const EMBED_DIM = 512;

async function computeEmbeddings() {
    // 动态导入新世界工厂
    const newWorldsModule = await import(pathToFileURL(path.join(_root, "src/new-worlds.js")).href);
    
    const worlds = [
        { name: "克苏鲁的呼唤", factory: newWorldsModule.createCthulhuWorld },
        { name: "都市怪谈·阈限空间", factory: newWorldsModule.createUrbanLegendWorld },
    ];

    console.log("Loading embedding model...");
    const extractor = await pipeline("feature-extraction", EMBED_MODEL);
    console.log("Model loaded.\n");

    for (const { name, factory } of worlds) {
        const world = factory();
        const snippets = world.lore_kb.snippets;
        console.log(`[${name}] Computing ${snippets.length} embeddings...`);

        for (let i = 0; i < snippets.length; i++) {
            const s = snippets[i];
            const text = [s.category, s.title, s.content, (s.keywords || []).join(" ")]
                .filter(Boolean).join(" ");
            const out = await extractor(text, { pooling: "mean", normalize: true });
            s.embedding = Array.from(out.data);
            s.embedDim = EMBED_DIM;
            s.embedModel = EMBED_MODEL;
            
            if ((i + 1) % 10 === 0 || i === snippets.length - 1) {
                console.log(`  ${i + 1}/${snippets.length}`);
            }
        }
    }

    // 输出嵌入结果 JSON
    const result = {};
    for (const { name, factory } of worlds) {
        const w = factory();
        result[w.id] = w.lore_kb.snippets.map(s => ({
            id: s.id,
            embedding: s.embedding,
            embedDim: s.embedDim,
            embedModel: s.embedModel
        }));
    }

    const outPath = path.join(_root, "data/new_world_embeddings.json");
    await writeFile(outPath, JSON.stringify(result, null, 2));
    console.log(`\n✓ Embeddings saved to ${outPath}`);
    
    // 打印统计
    for (const [worldId, snippets] of Object.entries(result)) {
        console.log(`  ${worldId}: ${snippets.length} snippets with ${(snippets[0]?.embedding?.length || 0)}-dim embeddings`);
    }
}

computeEmbeddings().catch(e => {
    console.error("Failed:", e);
    process.exit(1);
});

File: tools/compute_new_world_embeddings.py
# tools/compute_new_world_embeddings.py
# 为两个新世界计算向量嵌入（bge-small-zh-v1.5, 512维）
# 与游戏内 embed_model / embed_dim 一致
import json, sys, os
import numpy as np
from pathlib import Path

ROOT = Path(__file__).parent.parent
sys.path.insert(0, str(ROOT))

# 导入 sentence-transformers
try:
    from sentence_transformers import SentenceTransformer
except ImportError:
    print("需要安装 sentence-transformers: pip install sentence-transformers")
    sys.exit(1)

MODEL_NAME = "BAAI/bge-small-zh-v1.5"
EMBED_DIM = 512

print(f"Loading model: {MODEL_NAME}...")
model = SentenceTransformer(MODEL_NAME, trust_remote_code=True)

# 动态加载 new-worlds.js 不可能，改为直接嵌入 world 定义
# 我们从 new-worlds.js 手动提取 lore 条目
# 直接用 Node 生成带入参的 JSON 中间文件
print("Loading lore snippets from Node-generated JSON...")
input_path = ROOT / "data" / "new_world_snippets.json"
with open(input_path, "r", encoding="utf-8") as f:
    worlds_data = json.load(f)

for world_key, world in worlds_data.items():
    snippets = world["snippets"]
    print(f"\n[{world['name']}] Computing {len(snippets)} embeddings...")
    
    texts = []
    for s in snippets:
        text = " ".join(filter(None, [
            s.get("category", ""),
            s.get("title", ""),
            s.get("content", ""),
            " ".join(s.get("keywords", []))
        ]))
        texts.append(text)
    
    embeddings = model.encode(texts, normalize_embeddings=True)
    
    for s, emb in zip(snippets, embeddings):
        s["embedding"] = emb.tolist()
        s["embed_dim"] = EMBED_DIM
        s["embed_model"] = f"Xenova/{MODEL_NAME.split('/')[1]}"
    
    print(f"  Done. Embedding dim: {len(embeddings[0])}")

# 输出可注入的数据
output_path = ROOT / "data" / "new_world_embeddings.json"
# 仅保留嵌入元数据以便注入
result = {}
for world_key, world in worlds_data.items():
    result[world_key] = [
        {"id": s["id"], "embedding": s["embedding"], "embedDim": s["embed_dim"], "embedModel": s["embed_model"]}
        for s in world["snippets"]
    ]

with open(output_path, "w", encoding="utf-8") as f:
    json.dump(result, f, ensure_ascii=False, indent=2)

print(f"\nSaved to {output_path}")
for wk, snippets in result.items():
    print(f"  {wk}: {len(snippets)} snippets, {len(snippets[0]['embedding'])}-dim")
